<?xml version="1.0"?>

<!--This launch file will run all the necessary nodes to run the slam algorithm from a simulated robot. The simulation data will come from a csv file-->

<launch>
    
    <!--Define number of bots used in the simulation-->
    <arg name="num_of_bots" default="96"/>

    <arg name="csv_file" default="Example_96_bots.csv" />

    <arg name="min_datapoints" default="180"/>

    <!--     Ground Truth data-->
    <node name="ground_truth_pose" pkg="local_sensing_intest" type="simulated_apriltag_broadcaster.py" respawn="false"
          args="$(arg num_of_bots) $(arg csv_file)"/>

      <!--    Here the virutal TCP.launch file should start-->
    <node name="virtual_robot" pkg="local_sensing_intest" type="simulated_robot_tcp.py" args="$(arg num_of_bots) $(arg csv_file)" output="screen"
        respawn="true"/>

    <!--    Rviz-->
    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find local_sensing_intest)/rviz/config_file.rviz"
          respawn="true"/>

    <!--     Set initial pose using AprilTags, I have to change this to a custom initial pose from CSV, or create an apriltag simulator from CSV -->
    <!-- <node name="set_initial_pose" pkg="local_sensing_intest" type="initial_pose_from_apriltags.py" args="$(arg num_of_tags)" respawn="false"/> -->

    <rosparam command="load" file="$(find local_sensing_intest)/config/config.yml" />

    <!--     Transorms -->
    <include file="$(find local_sensing_intest)/launch/transforms_replay.launch">
        <arg name="num_of_bots" value="$(arg num_of_bots)" />
        <arg name="odom_est_type" value="rigid_joint" />
        <arg name="apply_corrections" value="false"/>
    </include>


    <!--Now we start other stuff needed for the slam algorithm to work-->>
    <!-- PointCloud -->
    <node name="point_cloud2_broadcaster" pkg="local_sensing_intest" type="point_cloud2_broadcaster_replay.py"
          respawn="true"
          output="screen" args="$(arg num_of_bots) $(arg min_datapoints)"/>
    <node name="pc_frame_transform" pkg="local_sensing_intest" type="pc_frame_transform.py"
          respawn="true" output="screen"/>

    <!-- PointCloud to LaserScan -->
    <include file="$(find local_sensing_intest)/launch/point_cloud_to_laserscan.launch"/>


    <!--    Feature extractor-->
    <node name="pc_feature_extractor" pkg="local_sensing_intest" type="pcl_feature_ext.py" respawn="true"
          output="screen"/>

    <!-- Gmapping -->
    <include file="$(find local_sensing_intest)/launch/gmapping.launch"/>

    <!-- Odometry -->
    <node name="odometry" pkg="local_sensing_intest" type="odometry.py" args="$(arg num_of_bots)" respawn="true" output="screen"/>

    <!--    IMU frame converter-->
    <!-- <node name="imu_frame_conversion" pkg="local_sensing_intest" type="imu_frame_conversion.py" respawn="true" args="$(arg num_of_bots)" -->
          <!-- output="screen"/> -->

    <!--    Save csv data-->
    <!-- <node type="csv_saver.py" name="csv_saver" pkg="local_sensing_intest" respawn="true" output="screen"/> -->
    

</launch>
